File: src/main/java/ru/hasa/springbootapp/service/UserServiceImpl.java
package ru.hasa.springbootapp.service;


import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import ru.hasa.springbootapp.dao.UserDao;
import ru.hasa.springbootapp.model.Role;
import ru.hasa.springbootapp.model.User;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class UserServiceImpl implements UserService, UserDetailsService {

    @Autowired
    private final UserDao userDao;

    public UserServiceImpl(UserDao userDao) {
        this.userDao = userDao;
    }

    @Override
    public void saveUser(User user) {
        userDao.saveUser(user);
    }

    @Override
    public void removeUserById(Long id) {
        userDao.removeUserById(id);
    }

    @Override
    public List<User> getAllUsers() {
        return userDao.getAllUsers();
    }

    @Override
    public User showUser(Long id) {
        return userDao.showUser(id);
    }

    @Override
    public User updateUser(User user) {
        return userDao.updateUser(user);
    }

    @Override
    public User findByUsername(String username) {
        return userDao.findByUsername(username);
    }

    @Override
    @Transactional
    public UserDetails loadUserByUsername(String mail) throws UsernameNotFoundException {
        User user = findByMail(mail);
        if (user == null) {
            throw new UsernameNotFoundException(String.format("Mail '%s' not found", mail));
        }

        return new org.springframework.security.core.userdetails.User(user.getMail(), user.getPassword(), mapRolesToAuthorities(user.getRoles()));
    }

    //Из коллекции ролей делаем коллекцию Authorities
    public Collection<? extends GrantedAuthority> mapRolesToAuthorities(Collection<Role> roles) {
        return roles.stream().map(r -> new SimpleGrantedAuthority(r.getName())).collect(Collectors.toList());
    }

    @Override
    public User findByMail(String mail) {
        return userDao.findByMail(mail);
    }
}

File: src/main/kotlin/ru/hasa/springbootapp/service/UserServiceImpl.kt
package ru.hasa.springbootapp.service

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ru.hasa.springbootapp.dao.UserDao
import ru.hasa.springbootapp.model.Role
import ru.hasa.springbootapp.model.User

@Service
class UserServiceImpl(private val userDao: UserDao) : UserService {

    override fun saveUser(user: User) {
        userDao.saveUser(user)
    }

    override fun removeUserById(id: Long) {
        userDao.removeUserById(id)
    }

    override fun getAllUsers(): List<User> = userDao.getAllUsers()

    override fun showUser(id: Long): User? = userDao.showUser(id)

    override fun updateUser(user: User): User = userDao.updateUser(user)

    override fun findByUsername(username: String): User? = userDao.findByUsername(username)

    @Transactional
    override fun loadUserByUsername(mail: String): UserDetails {
        val user = findByMail(mail) ?: throw UsernameNotFoundException(String.format("Mail '%s' not found", mail))
        return org.springframework.security.core.userdetails.User(user.mail, user.password, mapRolesToAuthorities(user.roles))
    }

    override fun mapRolesToAuthorities(roles: Collection<Role>): Collection<GrantedAuthority> =
        roles.map { r -> SimpleGrantedAuthority(r.name) }

    override fun findByMail(mail: String): User? = userDao.findByMail(mail)
}
